### ISSUE TITLE
VPA ControlledResources doesn't work.

### ISSUE THREAD

**Issue by nairb774** [kubernetes/autoscaler#3501]
Using a VPA like the following results in the CPU recommendations being produced and applied:

```yaml
apiVersion: autoscaling.k8s.io/v1
kind: VerticalPodAutoscaler
metadata:
  name: hamster
spec:
  targetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: hamster
  resourcePolicy:
    containerPolicies:
    - containerName: "*"
      controlledResources: ["memory"]
```

It looks like in the process of building the [container name to aggregate state map](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/recommender/routines/vpa.go#L27) the `ControlledResources` value is not copied from the working state to the per-container aggregate state in [AggregateStateByContainerName](https://github.com/kubernetes/autoscaler/blob/a80841eed5359df78c9bf806918c16af6a4aaf6d/vertical-pod-autoscaler/pkg/recommender/model/aggregate_container_state.go#L294-L306). This prevents the ControlledResource information from flowing down to where the filtering is applied.

[These changes](https://github.com/kubernetes/autoscaler/commit/eb72f989343e646ad41ba8f8f9ef48184c37e086), while hacky, address the problem and result in controlled resources working correctly. The main reason for not submitting a PR, is its not clear to me that this is the right change. In some ways the `ControlledResources` hanging off the `AggregateContainerState` breaks the abstraction boundary. Alternatively, [the filtering](https://github.com/tweks/autoscaler/blob/51eb724400a2ee366b3049b29a294ccf079c852d/vertical-pod-autoscaler/pkg/recommender/logic/recommender.go#L83-L85) is possibly being done at the wrong place in the pipeline. Maybe the ControlledResources should be observed in the updater or the admission controller rather than in the recommender?

@tweks you seem to have done the initial implementation - thoughts?

**Comment by rmb938** [comment]
I just ran into this in one of my clusters. I only want memory controlled but cpu ends up being recommended and applied as well.

**Comment by Rickyf115** [comment]
Ran into the same issue as well, cpu is recommended and applied even though memory is the only controlled resource.

**Comment by tweks** [comment]
Hey! Sorry for late reply, just came back from vacation. This was fixed by [this commit](https://github.com/kubernetes/autoscaler/pull/3118) in May, and will be released as part of VPA 0.9.

Re: recommender vs updater - we decided to filter is as early as possible (i.e. in recommender). The reason is that users would probably have good reason for VPA to control only one resource (eg. the other one is controlled by different system), and VPA's recommendation for the non-controlled resources could be misleading without this extra knowledge.

The common example is using HPA and VPA together (VPA controlling memory and HPA scaling on CPU). If HPA is configured with, let's say, target utilization 70%, then VPA will see the pods as underutilized and recommend smaller size (which is misleading, as HPA will keep pods utilized on 70% no matter their size).
